### ISSUE TITLE
Tpetra:  hang in graph->fillComplete

### ISSUE THREAD

**Issue by bathmatt** [trilinos/Trilinos#6838]
## Bug Report
@trilinos/tpetra 

@jhux2 @mhoemmen @rppawlo 

### Description
Running on 16 MPI ranks causes a critical problem to hang in a fill complete for EMPIRE

In this code in panzer in building the projection objects

```
Teuchos::RCP<const Tpetra::CrsGraph<LocalOrdinalT,GlobalOrdinalT,NodeT> >
BlockedTpetraLinearObjFactory<Traits,ScalarT,LocalOrdinalT,GlobalOrdinalT,NodeT>::
buildTpetraGraph(int i,int j) const
{
   using Teuchos::RCP;
   using Teuchos::rcp;

   // build the map and allocate the space for the graph and
   // grab the ghosted graph
   std::cout << "A "<<comm_->getRank()<<std::endl;
   RCP<const MapType> map_i = getMap(i);
   RCP<const MapType> map_j = getMap(j);
   std::cout << "B "<<comm_->getRank()<<std::endl;

   RCP<CrsGraphType> graph  = rcp(new CrsGraphType(map_i,0));
   RCP<const CrsGraphType> oGraph = getGhostedGraph(i,j);

   std::cout << "C "<<comm_->getRank()<<std::endl;
   // perform the communication to finish building graph
   RCP<const ExportType> exporter = getGhostedExport(i);
   std::cout << "D "<<comm_->getRank()<<std::endl;
   graph->doExport( *oGraph, *exporter, Tpetra::INSERT );
   std::cout << "E "<<comm_->getRank()<<std::endl;
   graph->fillComplete(map_j,map_i);
   std::cout << "F "<<comm_->getRank()<<std::endl;

   return graph;
}
```

 mpirun -np 16 /home/mbetten/build/debug/empire/EMPIRE-PIC/src/EMPIRE_PIC.exe --kokkos-numa=1 --kokkos-threads=1 --I=BDot.yaml

several times the code gets to and through this but after like 3 or 4 things we have all 16 ranks get to E and non get past that...  

### Steps to Reproduce
1. SHA1:  [insert here]
1. Configure script:  [attach here]
1. Configure log:  [attach here]
1. Build log:  [attach here]
1. Input deck:  [attach here]
1. Do this.
1. Do that.
1. Shake fist angrily at computer.
1. Run log:  [attach here]

**Comment by bathmatt** [comment]
commit 78a3577a86f5b32cfb31432d5e4ad4f3a2019b6f is our current testing branch, but testing head now

**Comment by bathmatt** [comment]
With head of 78a3577a86f5b32cfb31432d5e4ad4f3a2019b6f it now hangs on D and not E, so in the doExport call 
   graph->doExport( *oGraph, *exporter, Tpetra::INSERT );

everyone gets there, nobody gets past that

**Comment by mhoemmen** [comment]
@bathmatt Could you try rerunning with the following environment variable set?
```
TPETRA_VERBOSE=CrsGraph
```
This will dump a lot of output, but it's useful for tracking down hangs.

**Comment by cgcgcg** [comment]
@mhoemmen I believe this also happens on a release build, so it might not be the added checks.

**Comment by bathmatt** [comment]
It is a debug build and that file is attached
[out.err.gz](https://github.com/trilinos/Trilinos/files/4210860/out.err.gz)

**Comment by bathmatt** [comment]
It does happen in release as well..  I miss understood @cgcgcg comment

**Comment by bartlettroscoe** [comment]
Let's please make sure and add a test to the Tpetra test suite that shows this defect **before** this gets fixed.

**Comment by bathmatt** [comment]
I can get you a problem that shows this on the SRN, please email me on where/how to reproduce

**Comment by bartlettroscoe** [comment]
@kddevin (Trilinos Data Services Product Lead)

This has been identified as a blocker for EMPIRE where they can't get an update of Trilihos until this is fixed.  So from and ATDM standpoint, I think this would be considered very high priority.  Is that right @bathmatt?

But there are no failing Tpetra tests in any of the 37 ATDM Trilinos builds as shown yesterday [here](https://testing-dev.sandia.gov/cdash/index.php?project=Trilinos&date=2020-02-16&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=65&value1=Trilinos-atdm-&field2=subprojects&compare2=93&value2=Tpetra) so that means that Tpetra must be missing a test that covers the EMPIRE use case.  Therefore, [official Trilinos testing](https://github.com/trilinos/Trilinos/wiki/Trilinos-Testing-Policy) would say that a new failing test must be added to cover this use case (if this is indeed a defect) before the defect is fixed.

**Comment by bathmatt** [comment]
That is correct, it is a blocker as fixes were put into muelu which we are waiting to test.

It only happens on 16 MPI ranks, I ran 14 and it passed, 

Everyone goes into a tpetra code and it hangs, it happens every run and it just started

**Comment by bartlettroscoe** [comment]
> It only happens on 16 MPI ranks, I ran 14 and it passed,

Hum, not sure what to do about defects that only manifest on 16+ MPI ranks.  The Trilinos test suite does not have any tests that large.   Hopefully someone can construct a test case with less MPI ranks that will be able to reproduce the problem so a test like this can be added to the Trilinos test suite.

**Comment by bathmatt** [comment]
I think 1. we need to figure out what causes this, see if we can reproduce on 4 cpus, add test and fix.

**Comment by mhoemmen** [comment]
@bathmatt Could you please run again with `TPETRA_VERBOSE=1` ?  That will produce a much larger output file, but it could be helpful.

**Comment by bathmatt** [comment]
it will be some time after 11.  where can i dump this data for you?

**Comment by mhoemmen** [comment]
@bathmatt Some SRN testbed home directory perhaps?  It will probably be a few hundred MB.  Thanks!

**Comment by mhoemmen** [comment]
@bathmatt Some testbed home directory perhaps? It will probably be a few hundred MB. Thanks!

**Comment by bathmatt** [comment]
@mhoemmen it's in /tmp on vortex out.err.bz2

**Comment by bathmatt** [comment]
0x0000000006f5b98f in Tpetra::Details::countNumInCommon<__gnu_cxx::__normal_iterator<long long*, std::vector<long long, std::allocator<long long> > >, long long const*> (srcBeg=..., srcEnd=..., tgtBeg=0x10d20ab0, tgtEnd=0x10d20ae8)

Rank 8 is spinning here..


```
(gdb) where
#0  Tpetra::Details::countNumInCommon<__gnu_cxx::__normal_iterator<long long*, std::vector<long long, std::allocator<long long> > >, long long const*> (
    srcBeg=..., srcEnd=..., tgtBeg=0x10d20ab0, tgtEnd=0x10d20ae8)
    at /home/mbetten/Trilinos/Trilinos/packages/tpetra/core/src/Tpetra_Util.hpp:1011
#1  0x0000000006f57053 in Tpetra::Details::CrsPadding<int, long long>::merge_with_current_state (this=0x10d3b020, 
    phase=Tpetra::Details::CrsPadding<int, long long>::IMPORT, whichIndex=148, 
    tgtLclRowInd=294, tgtColInds=0x10d20ab0, numTgtEnt=7, 
    srcColInds=0xf0d74e0, numSrcEnt=5)
    at /home/mbetten/Trilinos/Trilinos/packages/tpetra/core/src/Tpetra_Details_CrsPadding.hpp:317
#2  0x0000000006f4e2a8 in Tpetra::Details::CrsPadding<int, long long>::update_impl (this=0x10d3b020, 
    phase=Tpetra::Details::CrsPadding<int, long long>::IMPORT, 
    whichImport=148, targetLocalIndex=294, tgtGblColInds=0x10d20ab0, 
    origNumTgtEnt=7, tgtIsUnique=false, srcGblColInds=0xf0d74e0, 
    origNumSrcEnt=5, srcIsUnique=false)
    at /home/mbetten/Trilinos/Trilinos/packages/tpetra/core/src/Tpetra_Details_CrsPadding.hpp:195
#3  0x0000000006f41239 in Tpetra::Details::CrsPadding<int, long long>::update_import (this=0x10d3b020, whichImport=148, targetLocalIndex=294, 
    tgtGblColInds=0x10d20ab0, origNumTgtEnt=7, tgtIsUnique=false, 


```

**Comment by mhoemmen** [comment]
@bathmatt Ah, I think I know what's going on.  I should have a fix for this in a day or two.

**Comment by bathmatt** [comment]
ok, please get me a patch asap so I can test before the MR process.  send a patch file to my SNL email please.
Thanks

**Comment by mhoemmen** [comment]
@bathmatt btw PR #6831 has these changes; it merged into develop 3 days ago.

**Comment by bathmatt** [comment]
what changes, the changes to cause the bug?  Not the fixes since it is still there.

**Comment by mhoemmen** [comment]
@bathmatt The changes to fix #6663, that fixed Sierra SD, and that might be causing you an issue (we don't know that it's a bug yet).

**Comment by mhoemmen** [comment]
@bartlettroscoe You're the best! :-D
